pyrobosim/pyrobosim/manipulation/grasping.py

```diff
@@ -0,0 +1,547 @@
+""" Grasping utilities. """
+
+from ..utils.pose import Pose
+
+from enum import Enum
+import numpy as np
+import matplotlib.pyplot as plt
+from mpl_toolkits.mplot3d import Axes3D
+from mpl_toolkits.mplot3d.art3d import Poly3DCollection
+from transforms3d.quaternions import rotate_vector, qinverse
+
+
+class GraspFace(Enum):
+    """ Enumerates grasp face types. """
+    UNKNOWN = 0
+    FRONT = 1
+    BACK = 2
+    TOP = 3
+    BOTTOM = 4
+    LEFT = 5
+    RIGHT = 6
+
+normal_from_face = {
+    GraspFace.UNKNOWN: None,
+    GraspFace.FRONT:   np.array([-1.0, 0.0, 0.0]),
+    GraspFace.BACK:    np.array([1.0, 0.0, 0.0]),
+    GraspFace.TOP:     np.array([0.0, 0.0, 1.0]),
+    GraspFace.BOTTOM:  np.array([0.0, 0.0, -1.0]),
+    GraspFace.LEFT:    np.array([0.0, 1.0, 0.0]),
+    GraspFace.RIGHT:   np.array([0.0, -1.0, 0.0]),
+}
+
+class GraspDirection(Enum):
+    """ Enumerates grasp direction types. """
+    UNKNOWN = 0
+    X_POS = 1
+    X_NEG = 2
+    Y_POS = 3
+    Y_NEG = 4
+    Z_POS = 5
+    Z_NEG = 6
+
+vec_from_direction = {
+    GraspDirection.UNKNOWN: None,
+    GraspDirection.X_POS:   np.array([1.0, 0.0, 0.0]),
+    GraspDirection.X_NEG:   np.array([-1.0, 0.0, 0.0]),
+    GraspDirection.Y_POS:   np.array([0.0, 1.0, 0.0]),
+    GraspDirection.Y_NEG:   np.array([0.0, -1.0, 0.0]),
+    GraspDirection.Z_POS:   np.array([0.0, 0.0, 1.0]),
+    GraspDirection.Z_NEG:   np.array([0.0, 0.0, -1.0]),
+}
```

Comment:
Not sure how this will then morph into supporting grasps at other angles. Maybe the direction should just be a unit vector, but the enum can help look up some common directions?

Reply:
Fitting a cylinder to the object would make this easier but I could imagine some objects (like a banana) where a cylinder doesn't work well...

Reply:
In an ideal world, we'd provide the option of doing cuboid, cylinder, sphere, etc.

While implementation was tough with the cuboid, I mostly picked it because it makes the number of samples for TAMP pretty small, assuming we're just targeting the center of each face as we do in this PR.

I know the MoveIt grasps package, for example, adds more complexity to this by
* Allowing corner grasps
* Sampling offsets from the center